op-e2e/sgt/sgt_test.go

```diff
@@ -0,0 +1,367 @@
+	require.Equal(t, common.Big0.Cmp(postSgtBalance), 0)
+	// balance invariant check
+	invariantBalanceCheck(t, ctx, sgt, testAddr, gasCost, txValue, depositSgtValue, depositL2Value, postSgtBalance)
+}
+
+func fullSGTGasPaymentAndNonZeroTxValueWithSufficientNativeBalanceSuccess(t *testing.T, ctx context.Context, index int64, sgt *SgtHelper) {
+	depositSgtValue := big.NewInt(10000000000000)
+	depositL2Value := big.NewInt(10000000000000)
+	txValue := big.NewInt(10000)
+	testAccount, testAddr, vaultBalanceBefore := setUpTestAccount(t, ctx, index, sgt, depositSgtValue, depositL2Value)
+
+	// make a simple tx with the testAccount: transfer txValue from testAccount to dummyAddr
+	tx, err := sgt.transferNativeToken(t, ctx, testAccount, dummyAddr, txValue)
+	require.NoError(t, err)
+	receipt, err := wait.ForReceiptOK(ctx, sgt.L2Client, tx.Hash())
+	require.NoError(t, err)
+	gasCost := calcGasFee(receipt)
+	vaultBalanceAfter := calcVaultBalance(t, ctx, sgt)
+
+	// gasCost == vaultBalanceDiff check
+	require.Equal(t, vaultBalanceAfter.Sub(vaultBalanceAfter, vaultBalanceBefore).Cmp(gasCost), 0)
+	// post sgt balance check: sgt should be used as gas first
+	opts := &bind.CallOpts{Context: ctx}
+	postSgtBalance, err := sgt.SgtContract.BalanceOf(opts, testAddr)
+	require.NoError(t, err)
+	require.Equal(t, new(big.Int).Add(postSgtBalance, gasCost).Cmp(depositSgtValue), 0)
+	// balance invariant check
+	invariantBalanceCheck(t, ctx, sgt, testAddr, gasCost, txValue, depositSgtValue, depositL2Value, postSgtBalance)
+}
+
+func partialSGTGasPaymentAndNonZeroTxValueWithSufficientNativeBalanceSuccess(t *testing.T, ctx context.Context, index int64, sgt *SgtHelper) {
+	depositSgtValue := big.NewInt(1000)
+	depositL2Value := big.NewInt(10000000000000)
+	txValue := big.NewInt(10000)
+	testAccount, testAddr, vaultBalanceBefore := setUpTestAccount(t, ctx, index, sgt, depositSgtValue, depositL2Value)
+
+	// make a simple tx with the testAccount: transfer txValue from testAccount to dummyAddr
+	tx, err := sgt.transferNativeToken(t, ctx, testAccount, dummyAddr, txValue)
+	require.NoError(t, err)
+	receipt, err := wait.ForReceiptOK(ctx, sgt.L2Client, tx.Hash())
+	require.NoError(t, err)
+	gasCost := calcGasFee(receipt)
+	vaultBalanceAfter := calcVaultBalance(t, ctx, sgt)
+
+	// gasCost == vaultBalanceDiff check
+	require.Equal(t, vaultBalanceAfter.Sub(vaultBalanceAfter, vaultBalanceBefore).Cmp(gasCost), 0)
+	// post sgt balance check: sgt should be used as gas first and should be spent all
+	opts := &bind.CallOpts{Context: ctx}
+	postSgtBalance, err := sgt.SgtContract.BalanceOf(opts, testAddr)
+	require.NoError(t, err)
+	require.Equal(t, common.Big0.Cmp(postSgtBalance), 0)
+	// balance invariant check
+	invariantBalanceCheck(t, ctx, sgt, testAddr, gasCost, txValue, depositSgtValue, depositL2Value, postSgtBalance)
+}
+
+func fullSGTInsufficientGasPaymentFail(t *testing.T, ctx context.Context, index int64, sgt *SgtHelper) {
+	depositSgtValue := big.NewInt(10000)
```

Comment:
Some comment about the choosing of `10000` will be helpful for future readers.

Reply:
good point, done in https://github.com/ethstorage/optimism/pull/125/commits/35c926638cfc153454310ba909ca159ae103e3aa